### PR TITLE
major updates to WebSockets implementation

### PR DESCRIPTION
**Pull request by analogrelay**
* Handshaking
* Ping/Pong handling
* UTF-8 validation
* Customization through WebSocketOptions
* Closes connection on Protocol Errors
* Autobahn Test Suite integrated

/cc @davidfowl @Tratcher @moozzyk @BrennanConroy

### CODE REVIEW

File: src/Microsoft.AspNetCore.WebSockets.Internal/HandshakeHelpers.cs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.WebSockets.Internal
+{
+    public static class HandshakeHelpers
+    {
+        /// <summary>
+        /// Gets request headers needed process the handshake on the server.
+        /// </summary>
+        public static readonly IEnumerable<string> NeededHeaders = new[]
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.WebSockets.Internal/HandshakeHelpers.cs]: Doesn't seem needed, no one is using this and its internal so we could probably get rid of it?

Comment (analogrelay) [reply src/Microsoft.AspNetCore.WebSockets.Internal/HandshakeHelpers.cs]: Quite likely. These were brought over from aspnet/WebSockets, so I'll do a pass to check for what's needed in there.

File: src/Microsoft.AspNetCore.WebSockets.Internal/HandshakeHelpers.cs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            if (!string.Equals("GET", request.Method, StringComparison.OrdinalIgnoreCase))
+            {
+                return false;
+            }
+
+            foreach (var pair in request.Headers)
+            {
+                if (string.Equals(Constants.Headers.Connection, pair.Key, StringComparison.OrdinalIgnoreCase))
+                {
+                    foreach(var value in pair.Value)
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.WebSockets.Internal/HandshakeHelpers.cs]: nit: formatting

File: src/Microsoft.AspNetCore.WebSockets.Internal/HandshakeHelpers.cs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+                    {
+                        if(string.Equals(Constants.Headers.ConnectionUpgrade, value, StringComparison.OrdinalIgnoreCase))
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.WebSockets.Internal/HandshakeHelpers.cs]: nit: formatting

File: src/Microsoft.AspNetCore.WebSockets.Internal/IHttpWebSocketConnectionFeature.cs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+﻿using System.Threading.Tasks;
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.WebSockets.Internal/IHttpWebSocketConnectionFeature.cs]: Copyright!

File: src/Microsoft.AspNetCore.WebSockets.Internal/Microsoft.AspNetCore.WebSockets.Internal.xproj [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+<Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <VisualStudioVersion Condition="'$(VisualStudioVersion)' == ''">14.0</VisualStudioVersion>
+    <VSToolsPath Condition="'$(VSToolsPath)' == ''">$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)</VSToolsPath>
+  </PropertyGroup>
+
+  <Import Project="$(VSToolsPath)\DotNet\Microsoft.DotNet.Props" Condition="'$(VSToolsPath)' != ''" />
+  <PropertyGroup Label="Globals">
+    <ProjectGuid>fffe71f8-e476-4bcd-9689-f106ee1c1497</ProjectGuid>
+    <RootNamespace>Microsoft.AspNetCore.WebSockets.Internal</RootNamespace>
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.WebSockets.Internal/Microsoft.AspNetCore.WebSockets.Internal.xproj]: Remove, maybe run Doug's script

File: src/Microsoft.AspNetCore.WebSockets.Internal/WebSocketAppBuilderExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+    public static class WebSocketAppBuilderExtensions
+    {
+        public static void UseWebSocketConnections(this IApplicationBuilder self, ChannelFactory channelFactory)
+        {
+            self.UseWebSocketConnections(channelFactory, new WebSocketConnectionOptions());
+        }
+
+        public static void UseWebSocketConnections(this IApplicationBuilder self, ChannelFactory channelFactory, WebSocketConnectionOptions options)
+        {
+            self.UseMiddleware<WebSocketConnectionMiddleware>(channelFactory, options);
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.WebSockets.Internal/WebSocketAppBuilderExtensions.cs:38]: Null checks?

File: src/Microsoft.AspNetCore.WebSockets.Internal/WebSocketConnectionMiddleware.cs [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+            _next = next;
+            _loggerFactory = loggerFactory;
+            _channelFactory = channelFactory;
+            _options = options;
+        }
+
+        public Task Invoke(HttpContext context)
+        {
+            var upgradeFeature = context.Features.Get<IHttpUpgradeFeature>();
+            if(upgradeFeature != null)
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.WebSockets.Internal/WebSocketConnectionMiddleware.cs]: nit: space

File: src/Microsoft.AspNetCore.WebSockets.Internal/WebSocketConnectionMiddleware.cs [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+            {
+                if(_options.ReplaceFeature || context.Features.Get<IHttpWebSocketConnectionFeature>() == null)
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.WebSockets.Internal/WebSocketConnectionMiddleware.cs]: nit: space

File: src/Microsoft.Extensions.WebSockets.Internal/Microsoft.Extensions.WebSockets.Internal.xproj [hunk 9]
```diff
@@ -4,18 +4,16 @@
     <VisualStudioVersion Condition="'$(VisualStudioVersion)' == ''">14.0</VisualStudioVersion>
     <VSToolsPath Condition="'$(VSToolsPath)' == ''">$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)</VSToolsPath>
   </PropertyGroup>
-
   <Import Project="$(VSToolsPath)\DotNet\Microsoft.DotNet.Props" Condition="'$(VSToolsPath)' != ''" />
   <PropertyGroup Label="Globals">
     <ProjectGuid>5d9da986-2eab-4c6d-bf15-9a4bdd4de775</ProjectGuid>
-    <RootNamespace>Microsoft.Extensions.WebSockets</RootNamespace>
+    <RootNamespace>Microsoft.Extensions.WebSockets.Internal</RootNamespace>
```
Comment (BrennanConroy) [inline src/Microsoft.Extensions.WebSockets.Internal/Microsoft.Extensions.WebSockets.Internal.xproj]: Remove

File: src/Microsoft.AspNetCore.WebSockets.Internal/IHttpWebSocketConnectionFeature.cs (hunk 4 above)
Comment (analogrelay) [reply src/Microsoft.AspNetCore.WebSockets.Internal/IHttpWebSocketConnectionFeature.cs]: Yep, I'll do a pass over all the files.

File: src/Microsoft.AspNetCore.Sockets/project.json [hunk 10]
```diff
@@ -1,13 +1,14 @@
...
+    "dependencies": {
+        "Channels": "0.2.0-beta-*",
+        "Microsoft.AspNetCore.Routing": "1.1.0-*",
+        "Microsoft.AspNetCore.WebSockets": "0.2.0-*",
+        "NETStandard.Library": "1.6.0"
     },
-    "net46": { }
-  }
+    "frameworks": {
+        "netstandard1.3": {
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Sockets/project.json]: nit: formatting

File: src/Microsoft.AspNetCore.WebSockets.Internal/WebSocketConnectionFeature.cs [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+        private async Task<IWebSocketConnection> AcceptWebSocketConnectionCoreAsync(string subProtocol)
+        {
+            _logger.LogDebug("Upgrading connection to WebSockets");
+            var opaqueTransport = await _upgradeFeature.UpgradeAsync();
+            var connection = new WebSocketConnection(
+                opaqueTransport.AsReadableChannel(),
+                _channelFactory.MakeWriteableChannel(opaqueTransport),
+                subProtocol: subProtocol);
+            _logger.LogDebug("Upgrading connection to WebSockets");
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.WebSockets.Internal/WebSocketConnectionFeature.cs]: Remove dupe log message

File: src/Microsoft.Extensions.WebSockets.Internal/WebSocketConnection.cs [hunk 12]
```diff
@@ -302,13 +401,116 @@ private void WriteMaskingKey(Span<byte> buffer)
...
+                        Dispose();
+                        return closeResult;
+                    }
+                    else if (!closed && _options.PassAllFramesThrough || (frame.Opcode != WebSocketOpcode.Ping && frame.Opcode != WebSocketOpcode.Pong))
+                    {
+                        await messageHandler(frame, state);
+                    }
+                }
+
+
```
Comment (BrennanConroy) [inline src/Microsoft.Extensions.WebSockets.Internal/WebSocketConnection.cs]: nit: newline

File: src/Microsoft.Extensions.WebSockets.Internal/WebSocketConnection.cs [hunk 13]
```diff
@@ -362,95 +642,67 @@ private Task SendCoreAsync(WebSocketFrame message, WebSocketCloseResult? closeRe
...
+            // Write the payload
+            payloadWriter(buffer, maskingKey, payloadLength, payload);
 
+            // Flush.
+            await buffer.FlushAsync();
 
-            // Commit and Flush
-            return buffer.FlushAsync();
+            // Unlock
+            _sendLock.Release();
```
Comment (BrennanConroy) [inline src/Microsoft.Extensions.WebSockets.Internal/WebSocketConnection.cs]: try .. finally?

File: src/Microsoft.AspNetCore.WebSockets.Internal/project.json [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+    "dependencies": {
+        "Microsoft.AspNetCore.Http.Abstractions": "1.1.0-*",
+        "Microsoft.Extensions.Logging.Abstractions": "1.1.0-*",
+        "Microsoft.Extensions.WebSockets.Internal": "0.1.0-*",
+        "NETStandard.Library": "1.6.1-*"
+    },
+
+    "frameworks": {
+        "net46": {},
+        "netstandard1.3": {
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.WebSockets.Internal/project.json:16]: nit: format

File: src/Microsoft.Extensions.WebSockets.Internal/project.json [hunk 15]
```diff
@@ -5,30 +5,15 @@
...
     "dependencies": {
         "Channels": "0.2.0-beta-*",
-        "Channels.Text.Primitives": "0.2.0-beta-*"
+        "Channels.Text.Primitives": "0.2.0-beta-*",
+        "NETStandard.Library": "1.6.0"
     },
 
     "frameworks": {
         "net46": {},
         "netstandard1.3": {
```
Comment (BrennanConroy) [inline src/Microsoft.Extensions.WebSockets.Internal/project.json]: nit: format

**Comment by davidfowl** [comment]
Need an update

File: src/Microsoft.AspNetCore.WebSockets.Internal/WebSocketAppBuilderExtensions.cs [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+
+using System;
+using Channels;
+using Microsoft.AspNetCore.WebSockets.Internal;
+
+namespace Microsoft.AspNetCore.Builder
+{
+    public static class WebSocketAppBuilderExtensions
+    {
+        public static void UseWebSocketConnections(this IApplicationBuilder self, ChannelFactory channelFactory)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.WebSockets.Internal/WebSocketAppBuilderExtensions.cs:18]: Add one that just makes a new channel factory

**Review by davidfowl — APPROVED** [review]
Few nits

File: src/Microsoft.AspNetCore.WebSockets.Internal/WebSocketAppBuilderExtensions.cs (hunk 16 above)
Comment (analogrelay) [reply src/Microsoft.AspNetCore.WebSockets.Internal/WebSocketAppBuilderExtensions.cs:18]: Ok. But who disposes it? The middleware?

**Pull request merged**